compiler/src/main/java/com/squareup/anvil/compiler/codegen/KotlinPoet.kt

```diff
@@ -62,13 +62,15 @@ internal fun ClassDescriptor.asClassName(): ClassName =
   )
 
 internal fun FqName.asClassName(module: ModuleDescriptor): ClassName {
-  try {
+  val segments = pathSegments().map { it.asString() }
+
+  // if the first sentence case is not the last segment it gets ambiguous:
+  // com.Foo.Bar could be a inner class Bar or an unconventional package com.Foo
+  val canGuessClassName = segments.indexOfFirst { it[0].isUpperCase() } == segments.size - 1
+  if (canGuessClassName) {
     return ClassName.bestGuess(asString())
-  } catch (ignored: IllegalArgumentException) {
-    // Probably lowercase class. Try to resolve the class below.
   }
 
-  val segments = pathSegments().map { it.asString() }
   for (index in (segments.size - 1) downTo 1) {
```

Comment:
`downTo 1` does not work well with the default package nested classes, but dagger compiler does not handle them either, so I did not bother fixing.